### PR TITLE
Add Haddocks for `updateField` and describe how to implement it. Refs #525.

### LINKED ISSUE #525: `copilot-core`: Add documentation for `updateField` describing how (and why) to implement it
[Copilot 3.20](https://github.com/Copilot-Language/copilot/releases/tag/v3.20) added support for struct updates via the new [`UpdateField`](https://hackage.haskell.org/package/copilot-core-3.20/docs/Copilot-Core-Operators.html#v:UpdateField) operation. One slightly non-obvious aspect of `UpdateField` is that `copilot-interpreter` doesn't support interpreting it unless you make sure to implement the [`updateField`](https://hackage.haskell.org/package/copilot-core-3.20/docs/Copilot-Core-Type.html#v:updateField) method in the `Struct` instance of any struct that whose field you update. For example, here is a naïve attempt at interpreting a Copilot specification involving a struct update:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE NoImplicitPrelude #-}
module Main (main) where

import Language.Copilot

data S = S
  { unS :: Field "unS" Word32
  }

instance Struct S where
  typeName _ = "s"
  toValues s = [Value typeOf (unS s)]

instance Typed S where
  typeOf = Struct (S (Field 0))

spec :: Spec
spec = do
  let externS :: Stream S
      externS = extern "extern_s" (Just [S (Field 0)])

      example :: Stream Word32
      example = (externS ## unS =: 42) # unS

  trigger "example" (example == example) [arg externS, arg example]

main :: IO ()
main = interpret 1 spec
```

This will fail with:

```
$ runghc UpdateFieldInterpreterFail.hs 
UpdateFieldInterpreterFail.hs: Field updates not supported for this type.
CallStack (from HasCallStack):
  error, called at src/Copilot/Core/Type.hs:64:17 in copilot-core-3.20-inplace:Copilot.Core.Type
```

To make this work, you need to implement `updateField` in the `Struct S` instance. Here is an example of how to do that:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Main (main) where

import Data.Proxy (Proxy(..))
import Data.Type.Equality (TestEquality(..), (:~:)(..))
import GHC.TypeLits (sameSymbol)
import qualified Prelude as P

import Language.Copilot

data S = S
  { unS :: Field "unS" Word32
  }

instance Struct S where
  typeName _ = "s"
  toValues s = [Value typeOf (unS s)]
  updateField s (Value fieldTy (field :: Field s a)) =
    case (sameSymbol (Proxy @s) (Proxy @"unS"), testEquality fieldTy Word32) of
      (Just Refl, Just Refl) -> s { unS = field }
      _ -> error $ "Unexpected field: " P.++ show field

instance Typed S where
  typeOf = Struct (S (Field 0))

spec :: Spec
spec = do
  let externS :: Stream S
      externS = extern "extern_s" (Just [S (Field 0)])

      example :: Stream Word32
      example = (externS ## unS =: 42) # unS

  trigger "example" (example == example) [arg externS, arg example]

main :: IO ()
main = interpret 1 spec
```
```
$ runghc UpdateFieldInterpreter.hs 
example:     
(<unS:0>,42)
```

The catch here is that I only knew how to do this because I was involved in the design of the `updateField` method. Unfortunately, there are currently no Haddocks for `updateField` itself, nor is there any kind of indication in the Copilot documentation that you would need to implement `updateField` like shown above in order to make `copilot-interpreter` work. Given how subtle this process is, I think there _should_ be some kind of tutorial in the documentation about how to do this.

It might also be helpful if the `Field updates not supported for this type` error message indicated that you need to implement `updateField`.

### PR DESCRIPTION
**Pull request by RyanGlScott**
This adds an `examples/StructsUpdateField.hs` example that (1) describes how to implement `updateField` for some example `Struct` instances, and (2) demonstrates that `updateField` works by using it with the Copilot interpreter. This also adds Haddocks to `updateField` that briefly describes how one might implement `updateField` in custom `Struct` instances. Because the full details of implementing `updateField` are somewhat intricate, these Haddocks refer to the `copilot/examples/StructsUpdateField.hs` example for further information.

Fixes #525.

### CODE REVIEW

**Comment by ivanperez-keera** [comment]
Change Manager: Please rebase on top of the current master.

**Comment by RyanGlScott** [comment]
I've rebased and moved the `updateField` example to its own file (`examples/StructsUpdateField.hs`). I've also left references to this example from the two existing struct-related examples (`examples/Structs.hs` and `examples/what4/Structs.hs`).

File: copilot/examples/Structs.hs [hunk 1]
```diff
@@ -23,6 +23,10 @@ instance Struct Volts where
   toValues volts = [ Value Word16 (numVolts volts)
                    , Value Bool   (flag volts)
                    ]
+  -- Note that we do not implement `updateField` here. `updateField` is only
+  -- needed to make the Copilot interpreter work, and we do not use the
```
Comment (ivanperez-keera) [inline copilot/examples/Structs.hs:27]: Could you rephrase this to "`updateField` is only needed to make updates to structs work in the Copilot interpreter?

Although it may be obvious to some, it sounds like without it structs just won't work at all with the interpreter.

Thanks!

**Comment by ivanperez-keera** [comment]
Change Manager: I'd like to make an additional change to a comment to make it a bit clearer. I left a note here: https://github.com/Copilot-Language/copilot/pull/547/commits/2a8916926763a7c39ac1fb90a7487b542605a5ad#r1807002785

File: copilot/examples/Structs.hs (hunk 1 above)
Comment (RyanGlScott) [reply copilot/examples/Structs.hs]: I've applied your suggested rephrasing.

**Review by ivanperez-keera — COMMENTED** [review]
Change Manager: Can you please modify [this commit message](https://github.com/Copilot-Language/copilot/pull/547/commits/d52e6f16cba9721dc7d591560150437833fac24f) to say that you also add more details in the error message printed to users if they have not implemented the method, to give them a hint as to why they cannot update values?

I'd also accept that change to the default implementation of `updateField` in a separate commit. In my mind, it's related but not an intricate part of the same change.

**Comment by RyanGlScott** [comment]
I've put those changes in a separate commit.

**Comment by ivanperez-keera** [comment]
Change Manager: Verified that:
- Solution is implemented:
  - [X] The code proposed compiles and passes all tests. Details:
    Build log: https://app.travis-ci.com/github/Copilot-Language/copilot/builds/272801025
  - [X] The solution proposed produces the expected result. Details:
    Evaluated by manual inspection. Running the following dockerfile checks that the new example and all examples modified continue running:
    ```Dockerfile
    FROM ubuntu:focal
    
    ENV DEBIAN_FRONTEND=noninteractive
    RUN apt-get update
    
    RUN apt-get install --yes libz-dev
    RUN apt-get install --yes git
    
    RUN apt-get install --yes wget
    RUN mkdir -p $HOME/.ghcup/bin
    RUN wget https://downloads.haskell.org/~ghcup/0.1.19.2/x86_64-linux-ghcup-0.1.19.2 -O $HOME/.ghcup/bin/ghcup
    
    RUN chmod a+x $HOME/.ghcup/bin/ghcup
    ENV PATH=$PATH:/root/.ghcup/bin/
    ENV PATH=$PATH:/root/.cabal/bin/
    RUN apt-get install --yes curl
    RUN apt-get install --yes gcc g++ make libgmp3-dev
    RUN apt-get install --yes pkg-config
    RUN apt-get install --yes z3
    
    SHELL ["/bin/bash", "-c"]
    
    RUN ghcup install ghc 9.4
    RUN ghcup install cabal 3.2
    RUN ghcup set ghc 9.4.8
    RUN cabal update
    
    CMD git clone $REPO \
        && cd $NAME \
        && git checkout $COMMIT \
        && cabal v1-sandbox init \
        && cabal v1-install alex happy --constraint='happy<2' \
        && cabal v1-install copilot**/ \
        && cabal v1-exec -- runhaskell copilot/examples/what4/Structs.hs \
        && cabal v1-exec -- runhaskell copilot/examples/Structs.hs \
        && cabal v1-exec -- runhaskell copilot/examples/StructsUpdateField.hs \
        && echo Success
    ```
    Command (substitute variables based on new path after merge):
    ```
    $ docker run -e "REPO=https://github.com/GaloisInc/copilot-1" -e "NAME=copilot-1" -e "COMMIT=75a3d9045dd6da368a2105ab8686129f71577075" -it copilot-verify-525
    ```
- [X] Implementation is documented. Details:
  The change is to the documentation of several modules and the examples. A new example is introduced and it is documented.
- [X] Change history is clear.
- [X] Commit messages are clear.
- [X] Changelogs are updated.
- [X] Examples are updated. Details:
  A new example is introduced, and other examples are updated to point to it as needed.
- [X] Required version bumps are evaluated. Details:
  No bump needed. Change is to the examples.
